New allocation docs (#68)
kubecost/docs: allocation.md

  -d shareIdle=false
  -G
```
```json
{
  "code": 200,
  "data": [
    {
      "__idle__": { ... },
      "default": { ... },
      "kube-system": { ... },
      "kubecost": { ... }
    },
    {
      "__idle__": { ... },
      "default": { ... },
      "kube-system": { ... },
      "kubecost": { ... }
    },
    {
      "__idle__": { ... },
      "default": { ... },
      "kube-system": { ... },
      "kubecost": { ... }
    },
    {
      "__idle__": { ... },
      "default": { ... },
      "kube-system": { ... },
      "kubecost": { ... }
    }
  ]
}
```

Note: querying for "3d" will likely return a range of four sets because the queried range will overlap with four precomputed 24-hour sets, each aligned to the configured timezone. For instance, querying "3d" on 2021/01/04T12:00:00 will return:
- 2021/01/04 00:00:00 until 2021/01/04T12:00:00 (now)
- 2021/01/03 00:00:00 until 2021/01/04 00:00:00
- 2021/01/02 00:00:00 until 2021/01/03 00:00:00
- 2021/01/01 00:00:00 until 2021/01/02 00:00:00

See [Querying](#querying) for the full list of arguments and [Examples](#examples) for more example queries.

## Allocation schema (version 1.76)
Field | Description
---: | :---

> name | Name of each relevant Kubernetes concept described by the allocation, delimited by slashes, e.g. "cluster/node/namespace/pod/container"

Reviewer: Maybe note that the schema differs by aggregation or even include the different schemas for each aggregation (though we may not want to commit to them).

Reviewer: I suppose `properties` should be the canonical location for this information.

Author: Right. As we've talked about, it might even be nice to deprecate name in favor is _just_ properties once the `Properties` struct is rewritten.

Author: This was probably the hardest one to define, for this reason.

Reviewer: Looks good as-is then. We don't have to commit to a schema if we're going to deprecate it soon-ish (hopefully :smile:).

properties | Map of name-to-value for all relevant property fields, including: `cluster`, `node`, `namespace`, `controller`, `controllerKind`, `pod`, `container`, `labels`, `annotation`, etc.

> window | Period of time over which the allocation is defined.

Reviewer: If this is the same as the window of the precomputed 24 hr sets mentioned earlier, mention that in this description because there's also the query parameter `window` with different meaning.

Reviewer: Actually, this is probably my own repeated confusion about `window`. I assume this is computed based on the `window=3d` query parameter, and then `start` and `end` are what I was talking about in the first comment. Could we then modify this definition to be more clear about what "over which the allocation is defined" means?

Something like: "Period of time (format: `FORMAT`) defined by the time of query and the `window` parameter. This will be the same for every object in `data[]`."

Author: "window" is one of the most loaded terms we have, so unfortunately it does get confusing. I like where you're headed with that suggestion, but even that isn't quite right because for "3d" you will get three different sets, the windows of which will sum to the full queried window, but it won't be the same for every object in `data[]`.

Reviewer: Oh, shoot. So `window` is actually the duration "precomputed 24 hr set" mentioned earlier, and then `start` and `end` are the actual bookends of the **allocation**, yeah?

Reviewer: Keeping this all straight is a bit tricky still :sweat_smile:

Author: How does this sound:
> Period of time (format: `"start": "2009-01-02T15:04:05Z07:00", "end": "2009-01-02T15:04:05Z07:00"`) within the queried window, over which the allocation is defined. Will be the same for each allocation in any given object within the `data` array.

Reviewer: Yeah, I think that sounds good. :+1:

start | Precise starting time of the allocation. By definition must be within the window.
end | Precise ending time of the allocation. By definition must be within the window.
minutes | Number of minutes running; i.e. the minutes from `start` until `end`.
cpuCores | Average number of CPU cores allocated while running.
cpuCoreRequestAverage | Average number of CPU cores requested while running.
cpuCoreUsageAverage | Average number of CPU cores used while running.
cpuCoreHours | Cumulative CPU core-hours allocated.
cpuCost | Cumulative cost of allocated CPU core-hours.

> cpuEfficiency | Fraction of CPU request that was used.

Reviewer: Is this "Average fraction of CPU request that was used while running"?

Author: If we want to use the word "average", it would be most accurate to say: "Ratio of `cpuCoreUsageAverage`-to-`cpuCoreRequestAverage`, meant to represent the fraction of requested resources that were used." Is that better, or too wordy?

Reviewer: I definitely prefer that rewording.
```suggestion
cpuEfficiency | Ratio of `cpuCoreUsageAverage`-to-`cpuCoreRequestAverage`, meant to represent the fraction of requested resources that were used.
```

gpuHours | Cumulative GPU-hours allocated.
gpuCost | Cumulative cost of allocated GPU-hours.
networkCost | Cumulative cost of network usage.
pvBytes | Average number of bytes of PersistentVolumes allocated while running.

> pvByteHours | Cumulative PersistenVolume byte-hours allocated.

Reviewer: ```suggestion
pvByteHours | Cumulative PersistentVolume byte-hours allocated.
```

Author: Goodness

pvCost | Cumulative cost of allocated PersistentVolume byte-hours.
ramBytes | Average number of RAM bytes allocated while running.
ramByteRequestAverage | Average number of RAM bytes allocated while running.
ramByteUsageAverage | Average number of RAM bytes used while running.
ramByteHours | Cumulative RAM byte-hours allocated.
ramCost | Cumulative cost of allocated RAM byte-hours.

> ramEfficiency | Fraction of RAM request that was used.

Reviewer: Likewise.

Author: See above

Reviewer: ```suggestion
ramEfficiency | Ratio of `ramByteUsageAverage`-to-`ramByteRequestAverage`, meant to represent the fraction of requested resources that were used.
```

sharedCost | Cumulative cost of shared resources, including: shared namespaces, shared labels, shared overhead.
externalCost | Cumulative cost of external resources.
totalCost | Total cumulative cost

> totalEfficiency | Cost-weighted average of `cpuEfficiency` and `ramEfficiency`

Reviewer: Should we explain what "cost-weighted average" means?

Author: I didn't want to get too technical w/r/t equations, but we could try to provide examples somewhere. What would you recommend we add here as a description? The equation would be `((cpuEfficiency * cpuCost) + (ramEfficiency * ramCost)) / (cpuCost + ramCost)`, which isn't terribly long.

Reviewer: I'm a fan of just including that equation. I think it can only help understanding.
```suggestion
totalEfficiency | Cost-weighted average of `cpuEfficiency` and `ramEfficiency`. In equation form: `((cpuEfficiency * cpuCost) + (ramEfficiency * ramCost)) / (cpuCost + ramCost)`
```

      "app_kubernetes_io_name": "cost-analyzer",
      "name": "kubecost",
      "pod_template_hash": "94dc86fc"
    },
    "namespace": "kubecost",
    "node": "gke-niko-pool-2-9182dfa7-okb2",
    "pod": "kubecost-cost-analyzer-94dc86fc-lwvrm",
    "service": [
      "kubecost-frontend",
      "kubecost-cost-analyzer"
    ]
  },
  "window": {
    "start": "2021-03-11T00:00:00-07:00",
    "end": "2021-03-12T00:00:00-07:00"
  },
  "start": "2021-03-11T07:00:00Z",
  "end": "2021-03-11T17:00:00Z",
  "minutes": 600,
  "cpuCores": 0.200399,
  "cpuCoreRequestAverage": 0.2,
  "cpuCoreUsageAverage": 0.004317,
  "cpuCoreHours": 2.00399,
  "cpuCost": 0.044344,
  "cpuEfficiency": 0.021583,
  "gpuHours": 0,
  "gpuCost": 0,
  "networkCost": 1.3e-05,
  "pvBytes": 11453246122.666668,
  "pvByteHours": 114532461226.66667,
  "pvCost": 0.005845,
  "ramBytes": 59791087.387687,
  "ramByteRequestAverage": 57671680,
  "ramByteUsageAverage": 52687168.319468,
  "ramByteHours": 597910873.876871,
  "ramCost": 0.001652,
  "ramEfficiency": 0.913571,
  "sharedCost": 0,
  "externalCost": 0,
  "totalCost": 0.051853,
  "totalEfficiency": 0.053611
}
```

### Special types of allocation

> - `__idle__` refers to resources on a cluster that were not dedicated to a Kubernetes object. An idle resource can be shared (proportionally or evenly) with the other allocations from the same cluster. (See the argument `shareIdle`.)

Reviewer: ```suggestion
- `__idle__` refers to resources on a cluster that were not dedicated to a Kubernetes object (e.g. unused CPU core-hours on a node). An idle resource can be shared (proportionally or evenly) with the other allocations from the same cluster. (See the argument `shareIdle`.)
```

> - `__unallocated__` refers to aggregated allocations without the selected `aggregate` field; e.g. aggregating by `label:app` might produce an `__unallocated__` allocation composed of allocation without the `app` label.

Reviewer: ```suggestion
- `__unallocated__` refers to aggregated allocations without the selected `aggregate` field; e.g. aggregating by `label:app` might produce an `__unallocated__` allocation composed of allocations without the `app` label.
```

- `__unmounted__` (or "Unmounted PVs") refers to the resources used by PersistentVolumes that aren't mounted to a Pod using a PVC, and thus cannot be allocated to a Pod.

## Querying

```
GET /model/allocation
```
Argument | Default | Description
--: | :--: | :--

> window (required) | — | Duration of time over which to query. Accepts: words like `today`, `week`, `month`, `yesterday`, `lastweek`, `lastmonth`; durations like `30m`, `12h`, `7d`; RFC3339 date pairs like `2021-01-02T15:04:05Z,2021-02-02T15:04:05Z`; unix timestamps like `1578002645,1580681045`.

Reviewer: ```suggestion
window (required) | — | Duration of time over which to query. Accepts: words like `today`, `week`, `month`, `yesterday`, `lastweek`, `lastmonth`; durations like `30m`, `12h`, `7d`; comma-separated RFC3339 date pairs like `2021-01-02T15:04:05Z,2021-02-02T15:04:05Z`; comma-separated unix timestamp (seconds) pairs like `1578002645,1580681045`.
```

aggregate | | Field by which to aggregate the results. Accepts: `cluster`, `namespace`, `controllerKind`, `controller`, `service`, `label:<name>`, and `annotation:<name>`. Also accepts comma-separated lists for multi-aggregation, like `namespace,label:app`.
accumulate | false | If `true`, sum the entire range of sets into a single set.
idle | true | If `true`, include idle cost (i.e. the cost of the un-allocated assets) as its own allocation. (See [special types of allocation](#special-types-of-allocation).)

> external | false | If `true`, include external costs in each allocation.

Reviewer: ```suggestion
external | false | If `true`, include [external costs](http://docs.kubecost.com/getting-started#out-of-cluster) in each allocation.
```

Is there a better link for OOC costs explanation than the install instructions?

Author: Not to my knowledge. Is this where you would send people, @dwbrown2?

Reviewer: Tagging @dwbrown2 again but I don't think we have to wait if he's busy. This link seems reasonable to me.
